### PR TITLE
[Serve] gate __del__ call behind hasattr check

### LINKED ISSUE #18772: [Serve] [Bug] Redeploying function causes "AttributeError: 'function' object has no attribute '__del__'"
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Serve

### What happened + What you expected to happen

Deploying a function twice gives the following traceback:
```
2021-09-20 15:59:17,902 INFO api.py:243 -- Updating deployment 'f'. component=serve deployment=f
(pid=87851) 2021-09-20 15:59:17,906     INFO backend_state.py:854 -- Stopping 1 replicas of deployment 'f' with outdated versions. component=serve deployment=f OLD VERSION: wLDnZI  NEW VERSION: nVoLLY
(pid=87853) 2021-09-20 15:59:19,909     ERROR backend_worker.py:360 -- Exception during graceful shutdown of replica. component=serve deployment=f replica=f#DrtNqg
(pid=87853) Traceback (most recent call last):
(pid=87853)   File "/Users/archit/ray/python/ray/serve/backend_worker.py", line 358, in prepare_for_shutdown
(pid=87853)     self.callable.__del__()
(pid=87853) AttributeError: 'function' object has no attribute '__del__'
(pid=87851) 2021-09-20 15:59:20,072     INFO backend_state.py:898 -- Adding 1 replicas to deployment 'f'. component=serve deployment=f
2021-09-20 15:59:20,400 INFO api.py:251 -- Deployment 'f' is ready at `http://127.0.0.1:8000/f`. component=serve deployment=f
```

### Reproduction script

```python
from ray import serve
@serve.deployment
def f(request):
    return "Hi"
serve.start()
f.deploy()
f.deploy()
```

### Anything else
Occurs on master
_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by architkulkarni**
## Related issue number

Closes #18772 
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/backend_worker.py [hunk 1]
```diff
@@ -355,7 +355,8 @@ async def prepare_for_shutdown(self):
         # We set the del method to noop after succssifully calling it so the
         # destructor is called only once.
         try:
-            self.callable.__del__()
+            if hasattr(self.callable, "__del__"):
```
Comment (architkulkarni) [inline python/ray/serve/backend_worker.py:358]: Not sure if it's better to put this outside the whole try-except block, but it's probably not very consequential

Comment (simon-mo) [reply python/ray/serve/backend_worker.py:358]: this should be totally fine!

**Review by simon-mo — APPROVED** [review]

**Comment by simon-mo** [comment]
note to self: this wasn't caught during that commit because the errors are silently logged but not thrown. In the future, we can add logging output test to capture any unexpected output.

**Comment by davidberenstein1957** [comment]
I was having the same issue. @simon-mo why isn't this set to `del self.callable`?

**Pull request merged**